### ISSUE TITLE
JSPM 2 cannot build Aurelia projects

### ISSUE THREAD

**Issue by KoltesDigital** [jspm/jspm-cli#2454]
Hi,

I'm trying to import Aurelia in a project, but it fails really early, because of `aurelia-bootstrapper` trying to detect the platform.

Using https://github.com/KoltesDigital/example-ts-jspm-aurelia/commit/2e27887b223ca91da9952149b70185126a7f25d0:

```bash
> jspm map ./dist/index.js ./dist/app.js -o importmap.json --map-base .
err  Error: Unable to resolve bare specifier "aurelia-loader-nodejs" from file:///C:/Projects/example-ts-jspm-aurelia/jspm_packages/npm/aurelia-bootstrapper@2.3.3/dist/commonjs/aurelia-bootstrapper.dew.js
    at throwBare (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\common.js:177:11)
    at Object.resolveImportMap (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\common.js:173:86)
    at MapResolver.mapResolve (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\index.js:233:58)
    at MapResolver.resolve (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\index.js:270:25)
    at MapResolver.resolveAll (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\index.js:241:31)
    at Promise.all.deps.map.dep (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\index.js:258:69)
    at Array.map (<anonymous>)
    at MapResolver.resolveAll (C:\Projects\example-ts-jspm-aurelia\node_modules\jspm\lib\map\index.js:258:53)
```

Original *aurelia-bootstrapper@2.3.3/dist/commonjs/aurelia-bootstrapper.js*:

```js
    if (isNodeLike && typeof module !== 'undefined' && typeof module.require !== 'undefined') 
    {
      var _m = module.require('aurelia-loader-nodejs');
      return Promise.resolve(new _m.NodeJsLoader());
    }
```

Rewritten *aurelia-bootstrapper@2.3.3/dist/commonjs/aurelia-bootstrapper.dew.js*:

```js
import { dew as _aureliaLoaderNodejsDew } from "aurelia-loader-nodejs";
...

      if (isNodeLike && true && typeof module.require !== 'undefined') {
        var _m = _aureliaLoaderNodejsDew();

        return Promise.resolve(new _m.NodeJsLoader());
      }
```

Whereas the original file loads dependencies on demand, the DEW loads them upfront. Of course I don't want to integrate the nodejs loader because the app targets the browser.

Do you think this problem can be solved on JSPM's side, or on Aurelia's side? (they have a custom bootstrapper for Webpack, so why not a custom bootstrapper for JSPM)

**Comment by guybedford** [comment]
Here's an override approach that can work for this:

```
jspm install aurelia-bootstrapper -o "map['aurelia-loader-nodejs']={browser:'@empty'}"
```

When I try that I get a new error though:
```
err  Error: Unable to resolve bare specifier "object" from file:///home/guybedford/Projects/jspm-cli/sandbox/jspm_packages/npm/aurelia-bootstrapper@2.3.3/dist/commonjs/aurelia-bootstrapper.dew.js
    at throwBare (/usr/lib/node_modules/jspm/lib/map/common.js:177:11)
    at Object.resolveImportMap (/usr/lib/node_modules/jspm/lib/map/common.js:173:86)
    at MapResolver.mapResolve (/usr/lib/node_modules/jspm/lib/map/index.js:233:58)
    at MapResolver.resolve (/usr/lib/node_modules/jspm/lib/map/index.js:270:25)
    at MapResolver.resolveAll (/usr/lib/node_modules/jspm/lib/map/index.js:241:31)
    at Promise.all.deps.map.dep (/usr/lib/node_modules/jspm/lib/map/index.js:258:69)
    at Array.map (<anonymous>)
    at MapResolver.resolveAll (/usr/lib/node_modules/jspm/lib/map/index.js:258:53)
```

Will see if I can find a fix for that too.

**Comment by guybedford** [comment]
I managed to trace the source of the issue above and release in jspm 2.0.0-beta.2, so the install as quoted there should get this to work for you.

**Comment by KoltesDigital** [comment]
Thanks! Indeed the importmap is generated.

However it now fails to load in the browser:

```
Uncaught (in promise) SyntaxError: The requested module 'blob:http://localhost:8080/c14cee7e-8104-43eb-a37a-306b3e60eb31' does not provide an export named 'dew'
async function (async)
topLevelLoad @ es-module-shims.js:713
(anonymous) @ es-module-shims.js:913
(anonymous) @ es-module-shims.js:934
```

Which still relates to `import { dew as _aureliaLoaderNodejsDew } from "aurelia-loader-nodejs";`. Maybe this will be a recurring pattern, so it may be worth adding it in @empty? In the meantime I wrote it:

*empty-dew.js*:

```js
export const dew = null;
```

```
jspm install aurelia-bootstrapper -o "map['aurelia-loader-nodejs']={browser:'empty-dew.js'}"
```

Then another problem is:

```
Uncaught (in promise) SyntaxError: The requested module 'blob:http://localhost:8080/38f0e029-13c2-4b96-8baf-ae4e9435d4c9' does not provide an export named 'bootstrap'
```

Due to the rewritten *commonjs/aurelia-bootstrapper.js* being only:

```js
import { dew } from './aurelia-bootstrapper.dew.js';
export default dew();
```

whereas the original file has many named exports. Although this should be the same at runtime, static analysis fails.

But at this point, I think the problems stem from looking in the *commonjs* directory, instead of the *es2015* one. In *aurelia-bootstrapper@2.3.3/package.json*, and therefore in *aurelia-bootstrapper@2.3.3/package.json.dew.js* there's:

```js
main: "dist/commonjs/aurelia-bootstrapper.js",
```

There's another file *dist/es2015/aurelia-bootstrapper.js* which just exports things fine, and also resolves `aurelia-loader-nodejs` in its place and not upfront. But I don't know how to override the resolution in order to point to this file, or how we are supposed to switch that.

Quick workaround in the meantime:

```js
import bootstrapper from "aurelia-bootstrapper";
bootstrapper.bootstrap(function(aurelia: Aurelia) { ...
```

Now another error:

```
Uncaught (in promise) No PLATFORM.Loader is defined and there is neither a System API (ES6) or a Require API (AMD) globally available to load your app.
```

The bottom line is `aurelia-pal` also suffers from importing from *commonjs* instead of *es2015*.

Can you do something for overriding the main file?

**Comment by guybedford** [comment]
Nicely done - `jspm install aurelia-bootstrapper -o "map['aurelia-loader-nodejs']={browser:'empty-dew.js'}"` is exactly the correct approach!

The second one is actually a bug in Aurelia in that it is expecting named exports to be defined for CommonJS modules, while the integration of ES modules in Node.js doesn't provide this (ie it won't work under `node --experimental-modules` either).

There is a way to cater to this via overrides as well - `jspm intall aurelia-bootstrapper -o "namedExports['commonjs/aurelia-bootstrapper.js']=[bootstrap]"`.

If you wanted to override the resolution to use the different main file then you could do - `jspm install aurelia-bootstrapper -o "main=dist/es2015/aurelia-bootstrapper.js"`.

The last bug you're seeing is actually a bug I believe I fixed the day after you posted this in https://github.com/jspm/jspm-cli/commit/af4e6097783641659f3b27f27d26ee8f7c5abe0e. So if you try with the latest jspm release you might have better luck!

**Comment by KoltesDigital** [comment]
Thanks!

Neither `jspm install aurelia-bootstrapper -o "namedExports['commonjs/aurelia-bootstrapper.js']=[bootstrap]" -o "map['aurelia-loader-nodejs']={browser:'empty-dew.js'}"` nor `jspm install aurelia-bootstrapper -o "namedExports['commonjs/aurelia-bootstrapper.js']=[bootstrap]; map['aurelia-loader-nodejs']={browser:'empty-dew.js'}"` does fix the export issue (`Uncaught (in promise) SyntaxError: The requested module 'blob:http://localhost:8080/4097a94a-9593-4c8b-bf24-8da187732217' does not provide an export named 'bootstrap'`).

Going with `jspm install aurelia-bootstrapper -o "main=dist/es2015/aurelia-bootstrapper.js"`, this creates a file *jspm_packages/npm/aurelia-bootstarpper@2.3.3/index.dew.js* containing `export { dew } from './dist/es2015/aurelia-bootstrapper.dew.js';`, but this latter file does not exist. If I stop there, the importmap is exported, but the browser complains

```
Uncaught (in promise) SyntaxError: The requested module 'blob:http://localhost:8080/2e46bc72-5e57-478a-a6bc-4835151d9acb' does not provide an export named 'PLATFORM'
```

which seems legit because *aurelia-pal* would also need to have the main file override. After running `jspm install aurelia-pal -o "main=dist/es2015/aurelia-pal.js"`, the same thing happen: *aurelia-pal@1.8.2/index.dew.js* imports the inexistent file *./dist/es2015/aurelia-pal.dew.js*. However this time, running `jspm map` fails:

```
err  ENOENT: no such file or directory, open 'G:\Projects\example-jspm-aurelia\jspm_packages\npm\aurelia-pal@1.8.2\dist\es2015\aurelia-pal.dew.js'
     Loading ./dist/es2015/aurelia-pal.dew.js from G:\Projects\example-jspm-aurelia\jspm_packages\npm\aurelia-pal@1.8.2\index.dew.js
```

Most *\*.js* files in the *dist* directory are accompanied with a *\*.dew.js* except in the directories *dist/es2015* and *dist/native-modules*. Is it normal?

**Comment by guybedford** [comment]
If you check the file at `jspm_packages/aurelia-bootstrapper@x.x.x/commonjs/aurelia-bootstrapper.js` does that file exist and contain the named exports you defined?

The `.dew` files are generated as part of the conversion of CommonJS into ES Modules and is exactly correct to see.

For packages that only have ES modules, you can add a `"type": "module"` or an `-o "type=module"` override and these packages won't then do the dew conversion.

It sounds like that might be a worthwhile way to get your ES2015 files running.

The reason why those invalid dew imports are there is because it is assuming this is a CommonJS package, and CommonJS can only load CommonJS, not ES modules. If you want your entry point to be an ES module you either need it to be ".mjs" or to have that `"type": "module"` in the package.json or override.

This is quite the treacherous path you're following :P happy to keep answering any further questions you have.

**Comment by guybedford** [comment]
To follow up - the simple mental model to bear in mind here is just the following: **if it works under `npm install && node --experimental-modules`, it will work in jspm and vice versa**, since we exactly follow the Node.js integration path.

Overrides can allow changing that, but the mental model of the above is what you should focus on. Getting Aurelia to support `node --experimental-modules` usage is thus the problem.

**Comment by KoltesDigital** [comment]
Agreed, I guess they have to do it with Node 12 at the end of the year. Maybe this search can help them.

But haha sly, do you mean there's a simpler way? :) I used jspm 0.17 for some projects, I really liked its simplicity in addition to SystemJS, and now that import maps are coming, I believe this is the way to go. (BTW, thanks for your gigantic work)

Overriding with `-o "type=module main=dist/es2015/aurelia-*.js` works and of course needs to be done on all imported Aurelia packages. Note that overriding `aurelia-loader-nodejs` is not needed anymore.

I made a new Loader, since Aurelia default browser loader supports either RequireJS or SystemJS.

But then, I give up on another error: `aurelia-templating` imports `Scope` from `aurelia-binding`, which is an interface, but `aurelia-templating`'s output files still import it even though they don't use it, so the browser complains

```
The requested module 'blob:http://localhost:8080/26c49693-e830-41d7-aac5-22eac1e49883' does not provide an export named 'Scope'
```

Nothing you can do here, I'll ask at Aurelia...

BTW, how the hot reload is supposed to work now?

For the record, beware of `aurelia-bootstrap` which relates to Twitter's Bootstrap and has nothing to do with `aurelia-bootstrapper`.

**Comment by KoltesDigital** [comment]
BTW, I see that SystemJS will support import maps under the hood. SystemJS has some useful additional features, like loader plugins for transpiling TypeScript, adding CSS, loading files as text, and hot module replacement. So I'm tempted to keep using it. In the guide you mention SystemJS as a legacy build mechanism. Can't we imagine using SystemJS like it was on 0.17, i.e. during development as well?

**Comment by guybedford** [comment]
@KoltesDigital this is a great point. I've been meaning to pick up the hot reloading development on SystemJS 2 again, in which case that is a benefit for the development workflow.

Would you say hot reloading is the main thing you'd be after here? Or is it the asset use cases?

For assets that are specified like CSS I'd be glad to integrate these into ES module shims as the spec work progresses.

For the asset use cases I think the best advice would be to use native ES module workflows.... ie `universalFetch(new URL('./file.txt'), import.meta.url)` sort of thing. Then to rather have Rollup support for ensuring that still works out in the build, which I would gladly do right away.

**Comment by KoltesDigital** [comment]
Hm, hard to compare the two features.

Modifying asset loading is great, but this can be generalized with an external process which watches fs changes. I especially think about the case where one source file is used to generate N other files server-side, and the later are used by the browser-side app. This can't easily be done with loader plugins. So a watcher process could overcome the lack of plugin support.

Whereas HMR relates more to the development comfort, even to the point that wasting quite some time to code the module state persistence is still a smaller drawback than the convenience of having instant changes. I'm frustrated when I have to reload the whole page to see the changes, especially if it requires some preparation steps (e.g. getting a token of some kind which is not saved in localStorage).

So IMHO, hot reloading is more important.

I agree for the universalFetch being somehow rewritten by Rollup could supersede the loader plugins. How do you envision that? Because you also need to trigger plugins when developing, i.e. Rollup is not used at this moment.

**Comment by guybedford** [comment]
As in the jspm 2 docs, I like the overall concept of two-phase builds for performance. Compile/transpile first, then have the native ES module graph with assets that gets optimized into a production-ready graph.

In this model the fetch to local asset files would be a fetch to the built files, and instead the local compilation process would handle the watching / updating / caching based on mtime of those.

During the build if multiple ".css" files are emitted, css concatentation and optimization can ocurr, but again if we have CSS imports for example this can be done in a way that is web-compatible as opposed to being some custom pipeline, which I really like as it gets us to having that optimization step all being mostly config free (except down to customizations). The default experience works though without config which is the win.

Ok I will prioritize getting some watch workflows going for SystemJS and Es Module Shims.

**Comment by KoltesDigital** [comment]
Agreed!

One thing though with that step, how would you manage the intermediary built files? I used to keep them in another gitignored folder, but now TypeScript and other IDE cleverness require a physical presence. I'm still not fan of merging generated files into an original source folder.

As I write this, I realize a possible way is to organize the project in two modules a la monorepo (lerna), the built files would be the actual main files of a module, and the application would be in another module. This would help to keep generated files away of other source files. I'll give it a try.

**Comment by guybedford** [comment]
> One thing though with that step, how would you manage the intermediary built files? I used to keep them in another gitignored folder, but now TypeScript and other IDE cleverness require a physical presence. I'm still not fan of merging generated files into an original source folder.

I like the model of "src" -> "lib" for the first-phase TypeScript build, followed by a build of "lib" -> "dist", and then "dist" is what is published.

Yes exactly, separating through separate packages could make this distinction clear too for tools.
